app/api/text-to-sql/shared.ts

```diff
@@ -0,0 +1,118 @@
+  is_nullable: z.string().transform((val) => val === "YES"),
+  data_type: z.string(),
+  table_schema: z.string(),
+  table_name: z.string(),
+});
+
+export const tableSchema = z.object({
+  name: z.string(),
+  columns: z.record(
+    z.string(),
+    columnSchema.extend({
+      foreignKey: relationshipSchema
+        .pick({
+          constraint_name: true,
+          foreign_table_schema: true,
+          foreign_table_name: true,
+          foreign_column_name: true,
+        })
+        .optional(),
+    })
+  ),
+});
+
+export const databaseSchema = z.record(z.string(), tableSchema);
+
+export const getDatabaseSchema = async () => {
+  const columnsResult = await db.execute(sql`SELECT
+    table_schema,
+    table_name,
+    column_name,
+    ordinal_position,
+    column_default,
+    is_nullable,
+    udt_name AS data_type -- 'udt_name' gives the underlying base type
+  FROM
+    information_schema.columns
+  WHERE
+    table_schema IN (${sql.join(Array.from(schemaFilter), sql`, `)})
+  ORDER BY
+    table_schema,
+    table_name,
+    ordinal_position;`);
+
+  const columns = z.array(columnSchema).safeParse(columnsResult.rows);
+  if (!columns.success) {
+    throw new Error("Failed to parse columns");
+  }
+
+  const relationshipsResult = await db.execute(sql`SELECT
+      con.constraint_name,
+      con.unique_constraint_schema AS foreign_table_schema,
+      rel.table_name AS foreign_table_name,
+      rel.column_name AS foreign_column_name,
+      kcu.table_schema,
+      kcu.table_name,
+      kcu.column_name
+  FROM
+      information_schema.referential_constraints AS con
+  JOIN
+      information_schema.key_column_usage AS kcu
+      ON con.constraint_name = kcu.constraint_name AND con.constraint_schema = kcu.table_schema
+  JOIN
+      information_schema.key_column_usage AS rel
+      ON con.unique_constraint_name = rel.constraint_name AND con.unique_constraint_schema = rel.table_schema
+  WHERE
+      kcu.table_schema IN (${sql.join(Array.from(schemaFilter), sql`, `)});`);
+
+  const relationships = z
+    .array(relationshipSchema)
+    .safeParse(relationshipsResult.rows);
+  if (!relationships.success) {
+    throw new Error("Failed to parse relationships");
+  }
+
+  const schema: z.infer<typeof databaseSchema> = {};
+  columns.data.forEach((column) => {
+    if (!schema[column.table_name]) {
+      schema[column.table_name] = {
+        name: column.table_name,
+        columns: {},
+      };
+    }
+    schema[column.table_name].columns[column.column_name] = column;
+  });
+  relationships.data.forEach((relationship) => {
+    if (
+      schema[relationship.table_name] &&
+      schema[relationship.table_name].columns[relationship.column_name]
+    ) {
+      schema[relationship.table_name].columns[relationship.column_name] = {
+        ...schema[relationship.table_name].columns[relationship.column_name],
+        foreignKey: relationship,
+      };
+    }
+  });
+
+  return schema;
```

Comment:
To some extent we're depending on a couple things in general with text to sql:
1. the schema having names that are understandable  semantically  e.g. not efgT2
2. the number of columns being reasonable - some data lakes can have ludicrous numbers of cols/rows

Reply:
Love the idea of just asking the db for the schema. I think it is possible although not common for their to be descriptions that describe the meanings of the cols. Those are looking more required in the era of sql agents.

Reply:
Exactly, same as the tool zod schemas, if it's not plain English readable then it's not going to work at all. I wonder though if you did have nonsense columns if it'd ask in the chat about what each column is, because I had it asking for some sample data when the `amount` field was accidentally a string!

The data lake part is interesting, I'm not too experienced in that area, how many columns are we talking about? Like a million?